var return type for proc doesn't work at c++ backend

### Example
`nim cpp -r -d:danger scratch.nim`
```nim
type
  MyType[T] = object
    a,b: T
  MyTypeAlias = MyType[float32]

var m: MyTypeAlias
proc identity(o: var MyTypeAlias): var MyTypeAlias = o
discard identity(m)
```

### Current Output
```
nim cpp -r -d:danger scratch.nim
Hint: used config file '/home/arne/proj/nim/Nim/config/nim.cfg' [Conf]
Hint: used config file '/home/arne/proj/nim/Nim/config/config.nims' [Conf]
Hint: used config file '/home/arne/.config/nim/nim.cfg' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: scratch [Processing]
CC: scratch.nim
/home/arne/.cache/nim/scratch_r/@mscratch.nim.cpp: In function 'tyObject_MyType__oV8Am5o7e8x2FpgusXVyUg* identity__vOCbByxSDV3Os2F1Z4o1Og(tyObject_MyType__oV8Am5o7e8x2FpgusXVyUg&)':
/home/arne/.cache/nim/scratch_r/@mscratch.nim.cpp:40:11: error: cannot convert 'tyObject_MyType__oV8Am5o7e8x2FpgusXVyUg' to 'tyObject_MyType__oV8Am5o7e8x2FpgusXVyUg*' in assignment
   40 |  result = o;
      |           ^
      |           |
      |           tyObject_MyType__oV8Am5o7e8x2FpgusXVyUg
Error: execution of an external compiler program 'g++ -c -w -w -fpermissive -O3 -fno-strict-aliasing -fno-ident -std=gnu++14 -funsigned-char  -I/home/arne/proj/nim/Nim/lib -I/tmp -o /home/arne/.cache/nim/scratch_r/@mscratch.nim.cpp.o /home/arne/.cache/nim/scratch_r/@mscratch.nim.cpp' failed with exit code: 1



Compilation exited abnormally with code 1
```

### Expected Output
nothing, it should just compile

* reduced:
```nim
proc identity(o: var int): var int = o
var m = 1
discard identity(m)
```

* semi-related:
this should probably be a CT error (`var result` was not initialized), although it might be impossible for cases involving branching with runtime values; so at least a runtime check should be used in debug mode
```nim
when true:
  proc identity(): var int = discard
  discard identity()
```

nim c: SIGSEGV (unless -d:danger)
nim cpp: SIGSEGV
```
/tmp/d45f/@mt10540.nim.cpp:49:6: error: declaration of reference variable 'result' requires an initializer
        NI& result;
            ^~~~~~
/tmp/d45f/@mt10540.nim.cpp:50:11: error: C-style cast from rvalue to reference type 'NI &' (aka 'long long &')
        result = (NI&)0;
```

* not sure why but that example cgen's in c++ as a return ref (`NI& result`), whereas the one in top example uses `NI*result`

* pointers are less safe than references but IMO are preferable for return type as they allow more flexible patterns, eg:
```nim
proc fun(a, b: var int): var int =
  if cond: a else: b
```
(IMO this should be legal, unless marked with {.noaliasing.}), and in anycase that's what's done with C backend

## workaround
```nim
when true:
  proc identity(o: var int): var int = o.addr[]
  var m = 1
  discard identity(m)
```

--- a/compiler/semexprs.nim
+++ b/compiler/semexprs.nim
@@ -1577,9 +1577,13 @@ proc takeImplicitAddr(c: PContext, n: PNode; isLent: bool): PNode =
         root.name.s, renderTree(n, {renderNoComments}), explanationsBaseUrl])
   case n.kind
   of nkHiddenAddr, nkAddr: return n
-  of nkHiddenDeref, nkDerefExpr: return n[0]
+  of nkDerefExpr: return n[0]
   of nkBracketExpr:
     if n.len == 1: return n[0]
+  of nkHiddenDeref:
+    # issue #13848
+    # `proc fun(a: var int): var int = a`
+    discard
   else: discard
   let valid = isAssignable(c, n, isLent)
   if valid != arLValue:

--- a/tests/varres/tvarres0.nim
+++ b/tests/varres/tvarres0.nim
@@ -32,26 +32,99 @@ echo getF().a
 
 block: # #13848
   template fun() =
-    var m = 1
-    var m2 = 10
+    block:
+      var m = 1
 
-    proc identity(o: var int): var int =
-      result = o
-      result += 5
+      proc identity(o: var int): var int =
+        result = o
+        result += 5
 
-    proc identity2(o: var int): var int =
-      result = m2
-      result += 100
+      identity(m) += 3
+      doAssert m == 5+4
 
-    identity(m) += 3
-    doAssert m == 5+4
+    block:
+      var m = 10
+      proc identity2(o: var int): var int =
+        result = m
+        result += 100
 
-    var ignored = 27
-    identity2(ignored) += 7
-    doAssert m2 == 10 + 100 + 7
+      var ignored = 27
+      identity2(ignored) += 7
+      doAssert m == 10 + 100 + 7
+
+    block:
+      iterator test3(o: var int): var int = yield o
+      var m = 1
+      for m2 in test3(m): m2+=3
+      doAssert m == 4
 
   static: fun()
   fun()
+
+  template fun2() =
+    block:
+      var m = 1
+      var m2 = 1
+      iterator test3(o: var int): (var int, var int) =
+        yield (o, m2)
+
+      for ti in test3(m):
+        ti[0]+=3
+        ti[1]+=4
+
+      doAssert (m, m2) == (4, 5)
+  fun2()
+  # static: fun2() # BUG: Error: attempt to access a nil address kind: rkInt
+
+  template fun3() =
+    block:
+      proc test4[T1](o: var T1): var int = o[1]
+      block:
+        var m = @[1,2]
+        test4(m) += 10
+        doAssert m[1] == 2+10
+      block:
+        var m = [1,2]
+        test4(m) += 10
+        doAssert m[1] == 2+10
+      block:
+        var m = (1, 2)
+        test4(m) += 10
+        doAssert m[1] == 2+10
+
+      proc test5[T1](o: var T1): var int = o.x
+      block:
+        type Foo = object
+          x: int
+        var m = Foo(x: 2)
+        test5(m) += 10
+        doAssert m.x == 2+10
+      block:
+        type Foo = ref object
+          x: int
+        var m = Foo(x: 2)
+        test5(m) += 10
+        doAssert m.x == 2+10
+
+      proc test6[T1](o: T1): var int = o.x
+      block:
+        type Foo = ref object
+          x: int
+        var m = Foo(x: 2)
+        test6(m) += 10
+        doAssert m.x == 2+10
+
+  fun3()
+  static: fun3()
+
+  when false:
+    # BUG:
+    # c: SIGSEGV
+    # cpp: error: call to implicitly-deleted default constructor of 'tyTuple__ILZebuYefUeQLAzY85QkHA'
+    proc test7[T](o: var T): (var int,) =
+      (o[1], )
+    var m = @[1,2]
+    test7(m)[0] += 10
 
 block:
   # example from #13848

--- a/tests/varres/tvarres0.nim
+++ b/tests/varres/tvarres0.nim
@@ -28,3 +28,38 @@ getF().a = 1234
 echo getF().a
 getF() = Foo(a: 12345)
 echo getF().a
+
+
+block: # #13848
+  template fun() =
+    var m = 1
+    var m2 = 10
+
+    proc identity(o: var int): var int =
+      result = o
+      result += 5
+
+    proc identity2(o: var int): var int =
+      result = m2
+      result += 100
+
+    identity(m) += 3
+    doAssert m == 5+4
+
+    var ignored = 27
+    identity2(ignored) += 7
+    doAssert m2 == 10 + 100 + 7
+
+  static: fun()
+  fun()
+
+block:
+  # example from #13848
+  type
+    MyType[T] = object
+      a,b: T
+    MyTypeAlias = MyType[float32]
+
+  var m: MyTypeAlias
+  proc identity(o: var MyTypeAlias): var MyTypeAlias = o
+  discard identity(m)

